workflow/workflow_test.go

```diff
@@ -1367,6 +1380,225 @@ func TestExpressionWithWhitespace(t *testing.T) {
+	t.Logf("MultiDependency workflow failed purposefully in %d ms", duration.Milliseconds())
+	var typedError *workflow.ErrNoMorePossibleOutputs
+	if !errors.As(err, &typedError) {
+		t.Fatalf("incorrect error type returned: %T", err)
+	}
+	assert.LessThan(t, duration.Seconds(), 4) // It will take 5 seconds if it fails to fail early.
+}
+
+var multiDependencyFailureWorkflowWithCancellation = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        fail_purposefully:
+          type:
+            type_id: bool
+steps:
+  canceled_step:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 0
+    stop_if: !expr '$.input.fail_purposefully'
+    deploy:
+      deployer_name: "test-impl"
+      deploy_time: 50 # 50 ms of delay to make the cancellation more reliable.
+  long_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 5000
+outputs:
+  workflow-success:
+    cancelled_step_output: !expr $.steps.canceled_step.outputs.success
+    long_wait_output: !expr $.steps.long_wait.outputs.success
+`
+
+func TestMultiDependencyWorkflowFailureWithCancellation(t *testing.T) {
+	// Tests failure when one dependency fails immediately, and the other one fails later.
+	preparedWorkflow := assert.NoErrorR[workflow.ExecutableWorkflow](t)(
+		getTestImplPreparedWorkflow(t, multiDependencyFailureWorkflowWithCancellation),
+	)
+	startTime := time.Now() // Right before execute to not include pre-processing time.
+	_, _, err := preparedWorkflow.Execute(context.Background(), map[string]any{
+		"fail_purposefully": true,
+	})
+	duration := time.Since(startTime)
+	assert.Error(t, err)
+	t.Logf("MultiDependency workflow failed purposefully in %d ms", duration.Milliseconds())
+	var typedError *workflow.ErrNoMorePossibleOutputs
+	if !errors.As(err, &typedError) {
+		t.Fatalf("incorrect error type returned: %T", err)
+	}
+	assert.LessThan(t, duration.Seconds(), 4) // It will take 5 seconds if it fails to fail early.
+}
+
+var multiDependencyFailureWorkflowWithErrorOut = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        fail_purposefully:
+          type:
+            type_id: bool
+steps:
+  disabled_step:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: hello
+    input:
+      fail: !expr $.input.fail_purposefully
+  long_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 7000
+outputs:
+  workflow-success:
+    disabled_output: !expr $.steps.disabled_step.outputs.success
+    long_wait_output: !expr $.steps.long_wait.outputs.success
+`
+
+func TestMultiDependencyWorkflowFailureWithErrorOut(t *testing.T) {
+	// Tests failure when one dependency fails immediately, and the other one fails later.
```

Comment:
OK, this is the same comment; again for a similar but distinct scenario.

Reply:
Fixed.